elastic/kibana
# [Index Management] Add dynamic tabs to the index details page

## Summary

Fixes https://github.com/elastic/kibana/issues/168699
This PR adds a new function to the extensions service in Index Management to register additional tabs into the index details page. I think we will be adding more content using this service and it might change the interface of the service later on.

To add a new tab, add Index Management as a dependency to your plugin and use the `addIndexDetailsTab` function in the extensions service. The interface for the tab is as follows:
```ts
export interface IndexDetailsTab {
  // a unique key to identify the tab
  id: IndexDetailsTabIds;
  // a text that is displayed on the tab label, usually a Formatted message component
  name: ReactNode;
  // a function that renders the content of the tab
  renderTabContent: (indexName: string, index: Index) => ReactNode;
  // a number to specify the order of the tabs
  order: number;
}
```

### How to test
1. Add a test tab from the ILM plugin using this test commit https://github.com/elastic/kibana/pull/169570/commits/74a771db65892c4a6f6f9ebab8d0a9382225079a
2. Start Kibana and ES with `yarn es snapshot` and `yarn start`
3. Navigate to Index Management and click on any index name
4. Verify that the tab is displayed correctly and the tab header is in the correct order

### 
Screenshot
<img width="1042" alt="Screenshot 2023-10-24 at 15 59 10" src="https://github.com/elastic/kibana/assets/6585477/7929f5db-8d2a-41d9-9e6d-4d1f74a93a35">


### Screen recording  

https://github.com/elastic/kibana/assets/6585477/6826f68c-c90d-47b1-ae3e-da6e6981d7ba


### Checklist



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
